compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirExternalDeclarationChecker.kt

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.lexer.KtTokens
+
+object FirExternalDeclarationChecker : FirBasicDeclarationChecker() {
+    override fun check(declaration: FirDeclaration, context: CheckerContext, reporter: DiagnosticReporter) {
+        val source = declaration.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+        val modifierList = with(FirModifierList) { source.getModifierList() }
+        val externalModifier = modifierList?.modifiers?.singleOrNull { it.token == KtTokens.EXTERNAL_KEYWORD } ?: return
+
+        if (declaration !is FirSimpleFunction) {
```

Comment:
There are more than this, e.g., anonymous function. I think the use of `FirFunction<*>` would be safer.

Reply:
That's why `FirSimpleFunction` is used here... `external` is _not_ allowed for constructors, property accessor, etc. If I used `FirFunction<*>` then there wouldn't be an error on `external constructor()` (see `testData/diagnostics/testsWithStdLib/native/nonFunction.fir.kt`).

Reply:
Oh, I see! Then, could you add some comments about that (for someone like me)? :)

Reply:
So it looks like using `FirFunction` is actually correct. I was mistaken about property accessors, `external` is [allowed](https://kotlinlang.org/docs/java-interop.html#using-jni-with-kotlin) for those. Therefore using `FirSimpleFunction` should be incorrect and the error _should_ show up for accessors (e.g., `external set`), but they were not. I dug into why and it looks like it's because we're [not generating/collecting diagnostics](https://github.com/JetBrains/kotlin/blame/70c98d34d205939fa259e986754810f6abb690b4/compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/collectors/AbstractDiagnosticCollector.kt#L165) for `FirDefaultPropertyAccessor`.

On the other hand, by changing it to `FirFunction` we will have no error for `external constructor`. Looks like in FE1.0 this is handled separately and will be included in what I mentioned in the PR description: "The rest of the checks (see ModifiersChecker in FE1.0) will be added later."

So what I'll do for now in this PR, is change it to `FirFunction` and add a comment that `WRONG_MODIFIER_TARGE`T for constructors will be handled separately. Then I'll separately fix the issue on not collecting diagnostics for `FirDefaultPropertyAccessor` which will become an issue for diagnostics on accessors such as `GETTER_VISIBILITY_DIFFERS_FROM_PROPERTY_VISIBILITY`.